src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs

```diff
@@ -0,0 +1,341 @@
+            var count = _excludeFilters.Count;
+            for (var i = 0; i < _excludeFilters.Count; i++)
+            {
+                if (_excludeFilters[i].IsTypeExcluded(metadata.UnderlyingOrModelType))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private ValidationState GetValidationEntry(object model)
+        {
+            if (model == null || _validationState == null)
+            {
+                return null;
+            }
+
+            ValidationState entry;
+            _validationState.TryGetValue(model, out entry);
+            return entry;
+        }
+
+        private struct Recursifier : IDisposable
+        {
+            private readonly ValidationVisitor _visitor;
+            private readonly object _container;
+            private readonly string _key;
+            private readonly ModelMetadata _metadata;
+            private readonly object _model;
+            private readonly object _newModel;
+            private readonly IValidationStrategy _strategy;
+
+            public static Recursifier DoTheNeedful(
+                ValidationVisitor visitor,
+                string key,
+                ModelMetadata metadata,
+                object model,
+                IValidationStrategy strategy)
+            {
+                var recursifier = new Recursifier(visitor, model);
+
+                visitor._container = visitor._model;
+                visitor._key = key;
+                visitor._metadata = metadata;
+                visitor._model = model;
+                visitor._strategy = strategy;
+
+                return recursifier;
+            }
+
+            public Recursifier(ValidationVisitor visitor, object newModel)
+            {
+                _visitor = visitor;
+                _newModel = newModel;
+
+                _container = _visitor._container;
+                _key = _visitor._key;
+                _metadata = _visitor._metadata;
+                _model = _visitor._model;
+                _strategy = _visitor._strategy;
+            }
+
+            public void Dispose()
+            {
+                _visitor._container = _container;
+                _visitor._key = _key;
+                _visitor._metadata = _metadata;
+                _visitor._model = _model;
+                _visitor._strategy = _strategy;
+
+                _visitor._currentPath.Remove(_newModel);
+            }
+        }
+
+        // Sorts validators based on whether or not they are 'required'. We want to run
+        // 'required' validators first so that we get the best possible error message.
+        private class ValidatorOrderComparer : IComparer<IModelValidator>
+        {
+            public static readonly ValidatorOrderComparer Instance = new ValidatorOrderComparer();
+
+            public int Compare(IModelValidator x, IModelValidator y)
+            {
+                var xScore = x.IsRequired ? 0 : 1;
```

Comment:
Could we add a `Order` property on validators instead? It'd be in line with the way we do ordering in other parts of the application.